### PR TITLE
feat: Add Range Visualizer for Raid Buff Major Ids [skip ci]

### PR DESCRIPTION
**Pull request by ben9964**
This MR adds range circles to players and yourself when taking major ids that give ranged abilities such as Guardian, Heart of the pack etc. It's pretty useful to know if your teammate takes guardian and see where you can stand to keep that benefit!

The way that it does this is by tracking the buffs that players take when in a raid by extending the already present chat listener in the RaidModel class. Each buff is kept track of and then the range visualizer will check if any of these are major ids that require a range to be displayed and display them.

The list of raid buffs and major ids that they give is hard coded in RaidKind.java for each of the raids.

### CODE REVIEW

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/models/raid/RaidModel.java]: This pattern is a bit too weak, although unlikely if someone sent a regular message of "player has chosen the X buff!" this will catch it. You should make it expect the raid chat banner and also keep the colours, as far as I'm aware all raids use the same colours for the buff messages so no need to ignore those.

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/models/raid/RaidModel.java]: As mentioned above, you shouldn't need to remove the colours and should be able to use `matches` not `find`

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/models/raid/RaidModel.java]: What if here, instead of just storing the name and buff name you calculated the major ID here. So pass the buff group, eg "Bedrock III" into `RaidBuffs` and return "Heart of the Pack"

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/models/raid/RaidModel.java]: if you did my above comment you should be able to just return `buffNames`

File: common/src/main/java/com/wynntils/models/raid/type/RaidBuffs.java
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/models/raid/type/RaidBuffs.java:1]: I think this information can be somewhat linked to `RaidKind` so that when you are in NOG, you only need to check for the relevant buffs

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: Noted, I can see how someone could trigger it by sending such a message. The reason I went the route of stripping the colours as you've mentionned in other comments is that it felt like features that rely on colours could become a chore to update if wynn just happens to decide they like pink more than orange for example. Happy to change it back though as originally I did have it this way.

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: I would be happy to change the structure such that we only track major Ids rather than every raid buff. The original logic was that they may be a useful feature in extending this tracking to support seeing the buffs your teammates have taken somewhere on screen as well as the buffs you've taken (similar to how lootrunning missions that you've taken are displayed)

File: common/src/main/java/com/wynntils/models/raid/type/RaidBuffs.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/type/RaidBuffs.java:1]: Would the static buff -> major id map being a part of the Enum for each RaidKind be sufficient? Would it be clean enough to have something like

```java
NEST_OF_THE_GROOTSLANGS(
            "Nest of the Grootslangs",
            StyledText.fromString("§2Nest of The Grootslangs"),
            Pattern.compile("^Slay the Restless$"),
            Map.of(
                "Beserk", Map.of(1, "Explosive Impact"),
                "Lightbearer", Map.of(3, "Transcendence"),
                ...
            )
)
```

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ShadowCat117) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: It's not too much of a bother. Wynn changes various things that we have to adapt for, so having to change the colour in a pattern is at least an easy fix

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ShadowCat117) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: Ah if that is something you want to do in the future then by all means keep it this way

File: common/src/main/java/com/wynntils/models/raid/type/RaidBuffs.java
Comment (ShadowCat117) [reply common/src/main/java/com/wynntils/models/raid/type/RaidBuffs.java:1]: I think that would be acceptable,

Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/type/RaidBuffs.java:1]: Resolved

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: Should be resolved now. I've removed the color Util file that I added and made the pattern a complete match with what chat says when gaining a buff (either hex code for others or yellow for yourself)

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: Resolved

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/models/raid/RaidModel.java]: If you change the pattern to `§#d6401eff(\uE009\uE002|\uE001) §#fa7f63ff(\w+)§#d6401eff has chosen the §#fa7f63ff(\w+ \w+)§#d6401eff buff!` then you can use `event.getOriginalStyledText().stripAlignment().getMatcher(RAID_CHOOSE_BUFF_PATTERN)` in the event to make it a little cleaner.

Comment (ShadowCat117) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: Also I will note here this pattern is not perfect, as nicknamed members will have the `§o` before their name and you will have to get their actual username from the hover event, see `RevealNicknamesFeature` on how you can do that

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/models/raid/RaidModel.java]: Leftover debug

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ShadowCat117) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: The yellow is from the chat mentions feature, by default it it the same for everyone

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: Removed

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: I'm now using `getOriginalStyledText()`, thanks for the heads up! its great to have the message as it was before wynntils touched it.

as for nicknames I've implemented what I think will work but I am not champion so I cannot nickname myself to test that unfortunately

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: Using original message removes the yellow (or other colour possibilities as i found out) as a problem

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ShadowCat117) [inline common/src/main/java/com/wynntils/models/raid/RaidModel.java]: This is not the right way to do this, we don't like features referencing each other like this.

I would suggest either add a new method to StyledTextUtils that will extract just the username or simply moving the `getNameAndNick` method to StyledTextUtils

Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: I will refactor this as soon as I'm home later today, I had a look at the architecture wiki page and couldn't see anything about not exposing parts of features publicly. Maybe I missed it though, a util function is perfect either way.

Comment (ShadowCat117) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: I don't know if the wiki is up to date, a lot has changed since it was originally made but the idea is that a feature should only rely on itself, any relevant Models/Services/Managers/Handlers and other util classes. It should never rely on something contained in another feature.

If 2 features are doing the same, or similar things like in this case, then it's best to move that logic somewhere else

Comment (Rafii2198) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: I don't think Wiki ever changed well except for addition of functions at some point, but the wiki is the same as it was when I joined which was during 1.18 days (there wasn't even a config screen at the time)

File: common/src/main/java/com/wynntils/features/chat/RevealNicknamesFeature.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/chat/RevealNicknamesFeature.java]: I'll spare sending the same message as @ShadowCat117 did in https://github.com/Wynntils/Wynntils/pull/2827/files#r1748046610. TLDR, move this method, but also don't forget to use it in this feature itself. Currently you are just "duplicating" the method, and using it in another feature which breaks feature abstraction.

The code itself also needs to be refined a bit. `StylexText#iterate` is meant to be used for changing parts of a `StyledText`, this case, you are ignoring it. You can use `StyledText` with for-each styled loops as it has `iterator` implemented.

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: Other than that. `playerName.startsWith("§o")` can be solved by an optional group in the regex, and a separate group for an unformatted name string.

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/raid/RaidModel.java]:
```suggestion
    public List<String> getMajorIdRaidBuffs(String playerName) {
```

I think this wording makes more sense.

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/raid/RaidModel.java]:
```suggestion
        if (!partyRaidBuffs.containsKey(playerName)) return List.of();

        List<String> buffNames = partyRaidBuffs.get(playerName);
        List<String> majorIds = new ArrayList<>();

        for (String buffName : buffNames) {
            String majorId = this.currentRaid.majorIdFromBuff(buffName);
            if (majorId == null) continue;

            majorIds.add(majorId);
        }

        return majorIds;
    }
```

I am not sure if this was an attempt of (premature) optimization, or you just coded it this way, but simple code is always better when performance is not an issue. 

You can optionally try to rewrite this using streams, so something like `playerRaidBuffs.stream().map(...)...`

File: common/src/main/java/com/wynntils/models/raid/type/RaidKind.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/raid/type/RaidKind.java]: Isn't it "Stonewalker"?

File: common/src/main/java/com/wynntils/models/raid/type/RaidKind.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/raid/type/RaidKind.java]: This will error if a new major id is added to the game, but not the list (so basically it make older versions of the mod not work). I suggest you move the tier parsing and error handling to the model itself, and just do a simple lookup here.

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: This method does return the list of Major Ids though not the list of Raid buffs

File: common/src/main/java/com/wynntils/features/chat/RevealNicknamesFeature.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/features/chat/RevealNicknamesFeature.java]: It's not possible to use the extracted util method as it is doing a different job from the one in `RevealNicknamesFeature`.

The one present there is performing an action for each hover part that *isnt* the nickname, while the util is just extracting the nick and username. I could obviously add a consumer to the method for a possible action to be done when dealing with parts of the styled text that aren't nickname parts, but this seems overcomplicated and messy.

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: I'm happy with the refactor you suggested and have refactored that. I did just write it that way.

If streams is the convention that you'd like to follow (as I saw in the range visualizer itself) then I can do so but I think that there's no benefit to using them. With such little data streams are slower and a simple for loop is usually more readable too.

File: common/src/main/java/com/wynntils/models/raid/type/RaidKind.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/type/RaidKind.java]: Resolved

File: common/src/main/java/com/wynntils/models/raid/type/RaidKind.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/type/RaidKind.java]: I don't see how it would error when a new major id is added to a buff or a new buff is added to the game? If the buff isn't present in that raid's list (e.g. Test II was added) then it will return null and the loop in the calling method will just continue to the next iteration as if it didn't find a major id from that buff at all. 

If a major id is added to an existing buff (e.g. Hollowed I begins giving the player Explosive Impact), then it will also return null and continue in the calling method's for loop as if the major id doesn't exist. I don't see where the error would be raised.

File: common/src/main/java/com/wynntils/features/chat/RevealNicknamesFeature.java
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/chat/RevealNicknamesFeature.java]: I still think it's possible to create some method that can find, parse and return "nickname" StyledTextParts. Although, you'd have to move it to a model now, as `StyledTextUtils` is reserved for special, but not feature specific utility methods. (It's hard to imagine the boundaries, but since you need a nickname regex, we should keep the code in a model).

File: common/src/main/java/com/wynntils/features/combat/RangeVisualizerFeature.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/combat/RangeVisualizerFeature.java:129]: Perhaps keep this a stream, make the code below also be a stream, concat them and then put them into a list. That'd keep the code style the same, and also shorten the code a tiny bit.

File: common/src/main/java/com/wynntils/features/combat/RangeVisualizerFeature.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/combat/RangeVisualizerFeature.java]:
```suggestion
    private Pair<CustomColor, Float> getCircleFromMajorId(String majorIdName) {
```

More conventional, although it's more of a nit.

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: `getRaidMajorIds`?

File: common/src/main/java/com/wynntils/models/raid/type/RaidKind.java
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/models/raid/type/RaidKind.java]: What if "Hollowed 2's Guardian" buff is moved to "Hollowed 3"? Basically, you check the first get, but not the second one.

File: common/src/main/java/com/wynntils/utils/mc/StyledTextUtils.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/utils/mc/StyledTextUtils.java:211]: I'll mark this for myself tomorrow. I may or may not want this in a model, rather than this util, as it's too specific to StyledText. It's likely that a model of some sort should hold this code.

Comment (ben9964) [reply common/src/main/java/com/wynntils/utils/mc/StyledTextUtils.java:211]: Fair enough I'll leave that to you then. I believe I've done what was necessary for this util function to work as expected.

With regard to the error if a major id was moved, if the major id was moved then the first level of the map in the code would still have Hollowed as a key but it would have 2 as the secondary map's key rather than 3 so the second get would return null and everything would continue as expected. calling get() on a map where the key isn't present simply returns null rather than raising an error. I believe a containsKey() check is implicit to the get call

File: common/src/main/java/com/wynntils/features/combat/RangeVisualizerFeature.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/features/combat/RangeVisualizerFeature.java]: Resolved

File: common/src/main/java/com/wynntils/models/raid/RaidModel.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/RaidModel.java]: Resolved

File: common/src/main/java/com/wynntils/models/raid/type/RaidKind.java
Comment (magicus) [reply common/src/main/java/com/wynntils/models/raid/type/RaidKind.java]: Also, we want to keep the `type` classes as free from logic as possible; ideally they should just be simple Records, Enums, Interfaces etc; just keeping data but not any "code". That should be handled by the corresponding Model.

File: common/src/main/java/com/wynntils/utils/mc/StyledTextUtils.java
Comment (magicus) [reply common/src/main/java/com/wynntils/utils/mc/StyledTextUtils.java:211]: Yeah, this could just live in RaidModel, right?

Comment (ShadowCat117) [reply common/src/main/java/com/wynntils/utils/mc/StyledTextUtils.java:211]: Although I can't think of any currently, there's probably other places where getting username is useful so RaidModel could work for now but not a permenant place for it imo

Comment (ben9964) [reply common/src/main/java/com/wynntils/utils/mc/StyledTextUtils.java:211]: It's definitely a util method rather than something that is exclusive to this Model or part of the game. Granted those use cases haven't presented themselves yet so I'm happy to move it to the RaidModel until that point if that's the consensus

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/utils/mc/StyledTextUtils.java:211]: Okay, let's keep this here for now.

File: common/src/main/java/com/wynntils/utils/mc/StyledTextUtils.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/utils/mc/StyledTextUtils.java:230]: I think in this case it's generally smarter to return null, as it makes error handling code nicer.

File: common/src/main/java/com/wynntils/models/raid/type/RaidKind.java
Comment (ben9964) [reply common/src/main/java/com/wynntils/models/raid/type/RaidKind.java]: Logic has been moved out of the Enum

**Review by kristofbolyai — APPROVED** [review]
Thanks for going through with all the reviews! This looks great now!

**Pull request merged**
